draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -2171,70 +2171,45 @@ See {{hash2curve-repo}} or {{WB19}}, Section 4.3 for details on implementing the
 
 ## Mappings for Montgomery curves {#montgomery}
 
-The mapping defined in {{elligator2}} implements Elligator 2 {{BHKL13}} for
-curves defined by the Weierstrass equation
-
-~~~
-    Y^2 = X^3 + C * X^2 + D * X
-~~~
-
-(Note that this equation is different from the one used in {{weierstrass}}.)
-This Weierstrass curve is equivalent to the Montgomery curve
+The mapping defined in this section applies to a target curve M defined by the equation
 
 ~~~
     K * t^2 = s^3 + J * s^2 + s
 ~~~
 
-by the following change of variables:
-
-- C = J / K
-- D = 1 / K^2
-- X = s / K
-- Y = t / K
-
-The Elligator 2 mapping given below returns a point (X, Y) on the
-Weierstrass curve defined above.
-This point can be converted to a point (s, t) on the equivalent
-Montgomery curve by computing
-
-- s = K * X
-- t = K * Y
-
-Note that when D == K == 1, the above two curve equations
-are identical and no conversion is necessary.
-This is the case, for example, for Curve25519 and Curve448 {{RFC7748}}.
-
 ### Elligator 2 Method {#elligator2}
 
-Preconditions: A Weierstrass curve Y^2 = X^3 + C * X^2 + D * X
-where C != 0, D != 0, and C^2 - 4 * D is non-zero and non-square in F.
+Preconditions: A Montgomery curve K * t^2 = s^3 + J * s^2 + s where
+J != 0, K != 0, and (J^2 - 4) / K^2 is non-zero and non-square in F.
 
 Constants:
 
-- C and D, the parameters of the elliptic curve.
+- J and K, the parameters of the elliptic curve.
 
 - Z, a non-square element of F.
   {{elligator-z-code}} gives a Sage {{SAGE}} script that outputs the RECOMMENDED Z.
 
-Sign of Y: Inputs u and -u give the same X-coordinate.
-Thus, we set sgn0(Y) == sgn0(u).
+Sign of t: Inputs u and -u give the same s-coordinate.
+Thus, we set sgn0(t) == sgn0(u).
 
 Exceptions: The exceptional case is Z * u^2 == -1, i.e., 1 + Z * u^2 == 0.
-Implementations must detect this case and set X1 = -C.
+Implementations must detect this case and set x1 = -(J / K).
 Note that this can only happen when q = 3 (mod 4).
 
 Operations:
 
 ~~~
-1.  X1 = -C * inv0(1 + Z * u^2)
-2.  If X1 == 0, set X1 = -C
-3. gX1 = X1^3 + C * X1^2 + D * X1
-4.  X2 = -X1 - C
-5. gX2 = X2^3 + C * X2^2 + D * X2
-6.  If is_square(gX1), set X = X1 and Y = sqrt(gX1)
-7.  Else set X = X2 and Y = sqrt(gX2)
-8.  If sgn0(u) != sgn0(Y), set Y = -Y
-9.  return (X, Y)
+1.  x1 = -(J / K) * inv0(1 + Z * u^2)
+2.  If x1 == 0, set x1 = -(J / K)
+3. gx1 = x1^3 + (J / K) * x1^2 + x1 / K^2
+4.  x2 = -x1 - (J / K)
+5. gx2 = x2^3 + (J / K) * x2^2 + x2 / K^2
+6.  If is_square(gx1), set x = x1 and y = sqrt(gx1)
+7.  Else set x = x2 and y = sqrt(gx2)
+8.   s = x * K
+9.   t = y * K
+10. If sgn0(u) != sgn0(t), set t = -t
```

Comment:
The way we define sign in the text, sign of t is equal to sign of u. Changing the sign of y and then computing t doesn't necessarily give the same result (depending on the value of K).

From the point of view of the document, it's much simpler to define in terms of t than in terms of y, because y is a purely internal variable to the function.

Note that this change to the document doesn't change signs for any existing suites, because all of those suites are for curves where K = 1.

Reply:
> From the point of view of the document, it's much simpler to define in terms of t than in terms of y, because y is a purely internal variable to the function.
Yes, of course.

> Changing the sign of y and then computing t doesn't necessarily give the same result (depending on the value of K).
Well, this is exactly my point. But anyway, there are no breaking changes since K=1 as you said. So we can proceed as is.